plugin/kprom/config.go

```diff
@@ -0,0 +1,79 @@
+package kprom
+
+import "github.com/prometheus/client_golang/prometheus"
+
+type cfg struct {
+	buckets                []float64 // Used for timings metrics
+	namespace              string
+	reg                    prometheus.Registerer
+	subsystem              string
+	includeDetailedE2EHist bool
+	includeThrottleHist    bool
+	includeSeedNodes       bool
+}
+
+func newCfg(opts ...Opt) cfg {
+	cfg := cfg{
+		buckets:                prometheus.ExponentialBuckets(0.001, 2, 12),
+		namespace:              "",
+		reg:                    prometheus.DefaultRegisterer,
+		subsystem:              "kafka",
+		includeDetailedE2EHist: false,
+		includeThrottleHist:    false,
+		includeSeedNodes:       false,
+	}
+	for _, opt := range opts {
+		opt.apply(&cfg)
+	}
+	return cfg
+}
+
+type Opt interface {
+	apply(*cfg)
+}
+
+type opt struct{ fn func(*cfg) }
+
+func (o opt) apply(c *cfg) { o.fn(c) }
+
+// WithRegisterer allows the user to set a registerer to be used with kprom,
+// if the opt is not set the default prometheus registry will be used
+func WithRegisterer(reg prometheus.Registerer) Opt {
+	return opt{func(c *cfg) { c.reg = reg }}
+}
+
+// WithNamespace allows the user to provide a namespace for metrics, defaults
+// to empty string
+func WithNamespace(ns string) Opt {
+	return opt{func(c *cfg) { c.namespace = ns }}
+}
+
+// WithSybsystem allows the user to provide a sybsystem name for metrics,
+// defaults to "kafka"
+func WithSubsystem(ss string) Opt {
+	return opt{func(c *cfg) { c.subsystem = ss }}
+}
+
+// WithBuckets allows the user to specify the buckets used for kprom
+// histrograms, defaults to
```

Comment:
histrograms => histograms

Also what do these buckets mean? It might be good to indicate 1ms, 2ms, 4ms, etc.
This is more detailed than prometheus.DefBuckets -- are the default buckets not really ... sufficient?

Reply:
The default buckets might be a good default, but I think for Kafka the latencies we expect and want to measure are often in a range that is lower then what the defaults are designed for. I pulled a couple of metrics from our services to illustrate. And this is e2e, the detailed e2e metrics will be lower. I know everybody is having different setups, but I think it makes sense to have buckets that provide more insight into the lower ranges. But I also see the argument to use the defaults and let users overwrite them. Maybe users should even be able to overwrite them per histogram.

```
P90
histogram_quantile(0.9, sum by (client_id, le) (rate(ns_kafka_e2e_node_latency_seconds_bucket[3m])))
{clientid="ex"}     5.94 ms
{clientid="ar"}     96.5 ms
{clientid="us"}     3.06 ms
{clientid="br"}     90.9 ms
{clientid="de"}     2.45 ms

P50
histogram_quantile(0.5, sum by (client_id, le) (rate(ns_kafka_e2e_node_latency_seconds_bucket[3m])))
{clientid="ex"}     2.33 ms
{clientid="ar"}     30.7 ms
{clientid="us"}     0.923 ms
{clientid="br"}     30.1 ms
{clientid="de"}     1.54 ms
```